File: moai-adk-ts/src/core/installer/phase-executor.ts
// @CODE:INST-002 |
// Related: @CODE:INST-002:API, @CODE:INIT-003:BACKUP

/**
 * @file Installation phase execution engine
 * @author MoAI Team
 */

import { execSync } from 'node:child_process';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { InstallationError } from '@/utils/errors';
import { isInsideMoAIPackage } from '@/utils/path-validator';
import { logger } from '@/utils/winston-logger';
import type { ContextManager } from './context-manager';
import { PhaseValidator } from './phase-validator';
import { ResourceInstaller } from './resource-installer';
import { TemplateProcessor } from './template-processor';
import type {
  InstallationConfig,
  InstallationContext,
  ProgressCallback,
} from './types';
import { executePhase } from './utils/phase-runner.js';
import {
  type BackupMetadata,
  saveBackupMetadata,
} from './backup-metadata';
import {
  hasAnyMoAIFiles,
  generateBackupDirName,
  getBackupTargets,
} from './backup-utils';

/**
 * Executes installation phases with dependency injection
 *
 * Responsibilities:
 * - Execute installation phases using template pattern
 * - Coordinate with ResourceInstaller and TemplateProcessor
 * - Track phase progress via ContextManager
 * - Delegate validation to PhaseValidator
 *
 * @tags @CODE:PHASE-EXECUTOR-001
 */
export class PhaseExecutor {
  private readonly contextManager: ContextManager;
  private readonly resourceInstaller: ResourceInstaller;
  private readonly templateProcessor: TemplateProcessor;
  private readonly validator: PhaseValidator;

  constructor(contextManager: ContextManager) {
    this.contextManager = contextManager;
    this.templateProcessor = new TemplateProcessor();
    this.resourceInstaller = new ResourceInstaller(this.templateProcessor);
    this.validator = new PhaseValidator();
  }

  /**
   * Execute preparation phase including backup creation
   */
  async executePreparationPhase(
    context: InstallationContext,
    progressCallback?: ProgressCallback
  ): Promise<void> {
    await executePhase(
      'preparation',
      'Phase 1: Preparation and backup...',
      context,
      this.contextManager,
      async () => {
        if (context.config.backupEnabled) {
          await this.createBackup(context.config);
        }
        await this.validator.validateSystemRequirements(context.config);
        return [];
      },
      progressCallback
    );
  }

  /**
   * Execute directory structure creation phase
   */
  async executeDirectoryPhase(
    context: InstallationContext,
    progressCallback?: ProgressCallback
  ): Promise<void> {
    await executePhase(
      'directory',
      'Phase 2: Creating directory structure...',
      context,
      this.contextManager,
      async () => {
        await this.createProjectDirectories(context.config);
        return [context.config.projectPath];
      },
      progressCallback
    );
  }

  /**
   * Execute resource installation phase
   */
  async executeResourcePhase(
    context: InstallationContext,
    progressCallback?: ProgressCallback
  ): Promise<void> {
    await executePhase(
      'resource',
      'Phase 3: Installing resources...',
      context,
      this.contextManager,
      async () => {
        const files: string[] = [];

        const claudeFiles = await this.resourceInstaller.installClaudeResources(
          context.config
        );
        files.push(...claudeFiles);

        const moaiFiles = await this.resourceInstaller.installMoaiResources(
          context.config
        );
        files.push(...moaiFiles);

        const memoryFile = await this.resourceInstaller.installProjectMemory(
          context.config
        );
        if (memoryFile) {
          files.push(memoryFile);
        }

        return files;
      },
      progressCallback
    );
  }

  /**
   * Execute configuration generation phase
   */
  async executeConfigurationPhase(
    context: InstallationContext,
    progressCallback?: ProgressCallback
  ): Promise<void> {
    await executePhase(
      'configuration',
      'Phase 4: Generating configurations...',
      context,
      this.contextManager,
      async () => {
        const files: string[] = [];

        // Generate Claude Code settings if needed
        const claudeSettingsPath = path.join(
          context.config.projectPath,
          '.claude',
          'settings.json'
        );

        if (!fs.existsSync(claudeSettingsPath)) {
          const settingsPath =
            await this.resourceInstaller.createClaudeSettings(context.config);
          if (settingsPath) {
            files.push(settingsPath);
          }
        } else {
          logger.debug('Claude settings already exists, skipping', {
            settingsPath: claudeSettingsPath,
            tag: 'DEBUG:CLAUDE-SETTINGS-EXISTS-001',
          });
          files.push(claudeSettingsPath);
        }

        // Generate MoAI configuration
        const moaiConfigPath = await this.resourceInstaller.createMoaiConfig(
          context.config
        );
        if (moaiConfigPath) {
          files.push(moaiConfigPath);
        }

        // Generate gitignore
        const gitignorePath = await this.resourceInstaller.createGitignore(
          context.config
        );
        if (gitignorePath) {
          files.push(gitignorePath);
        }

        return files;
      },
      progressCallback
    );
  }

  /**
   * Execute validation and finalization phase
   */
  async executeValidationPhase(
    context: InstallationContext,
    progressCallback?: ProgressCallback
  ): Promise<void> {
    await executePhase(
      'validation',
      'Phase 5: Validation and finalization...',
      context,
      this.contextManager,
      async () => {
        await this.validator.validateInstallation(context.config);
        await this.finalizeInstallation(context.config);
        return [];
      },
      progressCallback
    );
  }

  // ============================================================================
  // Phase Implementation Methods
  // ============================================================================

  /**
   * Create backup of existing project
   * @CODE:INIT-003:BACKUP - Phase A implementation (v0.2.1: Refactored with backup-utils)
   */
  private async createBackup(config: InstallationConfig): Promise<void> {
    if (isInsideMoAIPackage(config.projectPath)) {
      throw new InstallationError(
        'Cannot create backup inside MoAI-ADK package directory',
        {
          phase: 'BACKUP_IN_PACKAGE_DIR',
          projectPath: config.projectPath,
        }
      );
    }

    // v0.2.1: Check if ANY MoAI-ADK files exist (OR condition) - using backup-utils
    if (!hasAnyMoAIFiles(config.projectPath)) {
      logger.debug('No MoAI-ADK files found, skipping backup', {
        projectPath: config.projectPath,
        tag: 'INFO:SKIP-BACKUP-001',
      });
      return;
    }

    // Generate backup directory name - using backup-utils
    const backupDirName = generateBackupDirName();
    const backupPath = path.join(config.projectPath, backupDirName);

    // Create backup directory
    await fs.promises.mkdir(backupPath, { recursive: true });

    // v0.2.1: Get selective backup targets - using backup-utils
    const targets = getBackupTargets(config.projectPath);
    const backedUpFiles: string[] = [];

    // Backup files and directories
    for (const target of targets) {
      const srcPath = path.join(config.projectPath, target);
      const dstPath = path.join(backupPath, target);

      if (fs.statSync(srcPath).isDirectory()) {
        await this.templateProcessor.copyDirectory(srcPath, dstPath);
        backedUpFiles.push(`${target}/`);
      } else {
        await fs.promises.copyFile(srcPath, dstPath);
        backedUpFiles.push(target);
      }
    }

    logger.debug('Backup created', {
      backupPath,
      backedUpFiles,
      tag: 'SUCCESS:BACKUP-001',
    });

    // Save backup metadata (Phase A: SPEC-INIT-003 v0.2.1)
    const metadata: BackupMetadata = {
      timestamp: new Date().toISOString(),
      backup_path: backupDirName,
      backed_up_files: backedUpFiles,
      status: 'pending',
      created_by: 'moai init',
    };

    await saveBackupMetadata(config.projectPath, metadata);

    logger.info('Backup metadata saved', {
      metadataPath: path.join(
        config.projectPath,
        '.moai',
        'backups',
        'latest.json'
      ),
      backedUpFiles,
      tag: 'SUCCESS:BACKUP-METADATA-001',
    });
  }

  /**
   * Create project directory structure
   */
  private async createProjectDirectories(
    config: InstallationConfig
  ): Promise<void> {
    const directories = [
      config.projectPath,
      path.join(config.projectPath, '.claude'),
      path.join(config.projectPath, '.claude', 'logs'),
      path.join(config.projectPath, '.moai'),
      path.join(config.projectPath, '.moai', 'project'),
      path.join(config.projectPath, '.moai', 'specs'),
      path.join(config.projectPath, '.moai', 'reports'),
      path.join(config.projectPath, '.moai', 'memory'),
    ];

    for (const dir of directories) {
      await fs.promises.mkdir(dir, { recursive: true });
      logger.debug('Created directory', { dir, tag: 'DEBUG:DIR-CREATE-001' });
    }

    logger.debug('Project directories created', {
      tag: 'SUCCESS:CREATE-DIRECTORIES-001',
    });
  }

  /**
   * Finalize installation with post-install tasks
   */
  private async finalizeInstallation(
    config: InstallationConfig
  ): Promise<void> {
    // Set permissions on non-Windows
    if (process.platform !== 'win32') {
      const scriptsDir = path.join(config.projectPath, '.claude', 'hooks');
      if (fs.existsSync(scriptsDir)) {
        await this.setExecutablePermissions(scriptsDir);
      }
    }

    // Initialize Git if team mode
    if (config.mode === 'team') {
      await this.initializeGitRepository(config);
    }

    logger.debug('Installation finalized', {
      projectPath: config.projectPath,
      mode: config.mode,
      tag: 'SUCCESS:FINALIZE-INSTALLATION-001',
    });
  }

  // ============================================================================
  // Helper Methods
  // ============================================================================

  /**
   * Set executable permissions on scripts
   */
  private async setExecutablePermissions(scriptsDir: string): Promise<void> {
    const files = await fs.promises.readdir(scriptsDir);
    for (const file of files) {
      // Only .sh scripts need executable permissions now (Python hooks replaced with TS)
      if (file.endsWith('.sh')) {
        const filePath = path.join(scriptsDir, file);
        await fs.promises.chmod(filePath, 0o755);
      }
    }
  }

  /**
   * Initialize Git repository
   */
  private async initializeGitRepository(
    config: InstallationConfig
  ): Promise<void> {
    try {
      process.chdir(config.projectPath);
      execSync('git init', { stdio: 'ignore' });
      execSync('git add .', { stdio: 'ignore' });
      execSync('git commit -m "Initial MoAI-ADK project setup"', {
        stdio: 'ignore',
      });

      logger.debug('Git repository initialized', {
        tag: 'SUCCESS:INIT-GIT-001',
      });
    } catch (error) {
      logger.error('Git initialization failed', {
        error,
        tag: 'ERROR:INIT-GIT-001',
      });
    }
  }
}

File: src/moai_adk/core/project/phase_executor.py
# @CODE:INIT-003:PHASE | SPEC: .moai/specs/SPEC-INIT-003/spec.md | TEST: tests/unit/test_init_reinit.py
"""Phase 기반 설치 실행 모듈 (SPEC-INIT-003 v0.3.0)

5단계 Phase로 구성된 프로젝트 초기화 실행:
- Phase 1: Preparation (백업 생성: .moai/backups/{timestamp}/)
- Phase 2: Directory (디렉토리 구조 생성)
- Phase 3: Resource (템플릿 복사, 사용자 콘텐츠 보존)
- Phase 4: Configuration (설정 파일 생성)
- Phase 5: Validation (검증 및 마무리)
"""

import shutil
import subprocess
from collections.abc import Callable
from pathlib import Path

from rich.console import Console

from moai_adk.core.project.backup_utils import (
    generate_backup_dir_name,
    get_backup_targets,
    has_any_moai_files,
    is_protected_path,
)
from moai_adk.core.project.validator import ProjectValidator

console = Console()

# 진행상황 콜백 타입
ProgressCallback = Callable[[str, int, int], None]


class PhaseExecutor:
    """Phase 기반 설치 실행

    5단계 Phase:
    1. Preparation: 백업 및 시스템 검증
    2. Directory: 디렉토리 구조 생성
    3. Resource: 템플릿 리소스 복사
    4. Configuration: 설정 파일 생성
    5. Validation: 검증 및 마무리
    """

    # 필수 디렉토리 구조
    REQUIRED_DIRECTORIES = [
        ".moai/",
        ".moai/project/",
        ".moai/specs/",
        ".moai/reports/",
        ".moai/memory/",
        ".moai/backups/",
        ".claude/",
        ".claude/logs/",
    ]

    def __init__(self, validator: ProjectValidator) -> None:
        """초기화

        Args:
            validator: 검증 인스턴스
        """
        self.validator = validator
        self.total_phases = 5
        self.current_phase = 0

    def execute_preparation_phase(
        self,
        project_path: Path,
        backup_enabled: bool = True,
        progress_callback: ProgressCallback | None = None,
    ) -> None:
        """Phase 1: 준비 및 백업

        Args:
            project_path: 프로젝트 경로
            backup_enabled: 백업 활성화 여부
            progress_callback: 진행상황 콜백
        """
        self.current_phase = 1
        self._report_progress(
            "Phase 1: Preparation and backup...", progress_callback
        )

        # 시스템 요구사항 검증
        self.validator.validate_system_requirements()

        # 프로젝트 경로 검증
        self.validator.validate_project_path(project_path)

        # 백업 생성
        if backup_enabled and has_any_moai_files(project_path):
            self._create_backup(project_path)

    def execute_directory_phase(
        self,
        project_path: Path,
        progress_callback: ProgressCallback | None = None,
    ) -> None:
        """Phase 2: 디렉토리 생성

        Args:
            project_path: 프로젝트 경로
            progress_callback: 진행상황 콜백
        """
        self.current_phase = 2
        self._report_progress(
            "Phase 2: Creating directory structure...", progress_callback
        )

        for directory in self.REQUIRED_DIRECTORIES:
            dir_path = project_path / directory
            dir_path.mkdir(parents=True, exist_ok=True)

    def execute_resource_phase(
        self,
        project_path: Path,
        progress_callback: ProgressCallback | None = None,
    ) -> list[str]:
        """Phase 3: 리소스 설치

        Args:
            project_path: 프로젝트 경로
            progress_callback: 진행상황 콜백

        Returns:
            생성된 파일 목록
        """
        self.current_phase = 3
        self._report_progress(
            "Phase 3: Installing resources...", progress_callback
        )

        # TemplateProcessor를 통해 리소스 복사 (silent 모드)
        from moai_adk.core.template import TemplateProcessor

        processor = TemplateProcessor(project_path)
        processor.copy_templates(backup=False, silent=True)  # Progress Bar 충돌 방지

        # 생성된 파일 목록 반환 (간소화)
        return [
            ".claude/",
            ".moai/",
            "CLAUDE.md",
            ".gitignore",
        ]

    def execute_configuration_phase(
        self,
        project_path: Path,
        config: dict[str, str],
        progress_callback: ProgressCallback | None = None,
    ) -> list[str]:
        """Phase 4: 설정 생성

        Args:
            project_path: 프로젝트 경로
            config: 설정 딕셔너리
            progress_callback: 진행상황 콜백

        Returns:
            생성된 파일 목록
        """
        self.current_phase = 4
        self._report_progress(
            "Phase 4: Generating configurations...", progress_callback
        )

        import json
        from moai_adk import __version__

        # config에 버전 정보 추가 (v0.3.1+)
        config["moai_adk_version"] = __version__
        config["optimized"] = False  # 초기값

        # config.json 생성
        config_path = project_path / ".moai" / "config.json"
        with open(config_path, "w", encoding="utf-8") as f:
            json.dump(config, f, indent=2, ensure_ascii=False)

        return [str(config_path)]

    def execute_validation_phase(
        self,
        project_path: Path,
        mode: str = "personal",
        progress_callback: ProgressCallback | None = None,
    ) -> None:
        """Phase 5: 검증 및 마무리

        Args:
            project_path: 프로젝트 경로
            mode: 프로젝트 모드 (personal/team)
            progress_callback: 진행상황 콜백
        """
        self.current_phase = 5
        self._report_progress(
            "Phase 5: Validation and finalization...", progress_callback
        )

        # 설치 결과 검증
        self.validator.validate_installation(project_path)

        # Team 모드: Git 초기화
        if mode == "team":
            self._initialize_git(project_path)

    def _create_backup(self, project_path: Path) -> None:
        """백업 생성 (Selective) - v0.3.0

        Args:
            project_path: 프로젝트 경로
        """
        # 백업 디렉토리 생성 (v0.3.0: .moai/backups/{timestamp}/)
        timestamp = generate_backup_dir_name()
        backup_path = project_path / ".moai" / "backups" / timestamp
        backup_path.mkdir(parents=True, exist_ok=True)

        # 백업 대상 가져오기
        targets = get_backup_targets(project_path)
        backed_up_files: list[str] = []

        # 백업 실행
        for target in targets:
            src_path = project_path / target
            dst_path = backup_path / target

            if src_path.is_dir():
                self._copy_directory_selective(src_path, dst_path)
                backed_up_files.append(f"{target}/")
            else:
                dst_path.parent.mkdir(parents=True, exist_ok=True)
                shutil.copy2(src_path, dst_path)
                backed_up_files.append(target)

        # Progress Bar 충돌 방지를 위해 메시지 제거
        # (Phase 1 진행상황 메시지로 충분)

    def _copy_directory_selective(self, src: Path, dst: Path) -> None:
        """보호 경로를 제외하고 디렉토리 복사

        Args:
            src: 소스 디렉토리
            dst: 대상 디렉토리
        """
        dst.mkdir(parents=True, exist_ok=True)

        for item in src.rglob("*"):
            rel_path = item.relative_to(src)

            # 보호 경로 제외
            if is_protected_path(rel_path):
                continue

            dst_item = dst / rel_path
            if item.is_file():
                dst_item.parent.mkdir(parents=True, exist_ok=True)
                shutil.copy2(item, dst_item)
            elif item.is_dir():
                dst_item.mkdir(parents=True, exist_ok=True)

    def _initialize_git(self, project_path: Path) -> None:
        """Git 저장소 초기화

        Args:
            project_path: 프로젝트 경로
        """
        try:
            subprocess.run(
                ["git", "init"],
                cwd=project_path,
                check=True,
                capture_output=True,
            )
            # Progress Bar 충돌 방지를 위해 메시지 제거
        except subprocess.CalledProcessError:
            # 에러 발생 시에만 로깅 (실패해도 치명적이지 않음)
            pass

    def _report_progress(
        self, message: str, callback: ProgressCallback | None
    ) -> None:
        """진행상황 보고

        Args:
            message: 진행 메시지
            callback: 콜백 함수
        """
        if callback:
            callback(message, self.current_phase, self.total_phases)
        # callback이 없으면 Progress Bar가 없다는 뜻이므로 출력하지 않음
        # (CLI 외부에서 사용 시에는 result만 확인하면 됨)
